### ISSUE TITLE
Re-add old redirect strategy

### ISSUE THREAD

**Issue by rymndhng** [dakrone/clj-http#361]
Since the redirect-strategy has been pushed into ApacheHttpClient, we've noticed several behavioural changes, i.e. #345 and #326 

The code is confusing to understand because the clojure redirect-code doesn't actually get executed (instead we see exceptions when the max redirect is encountered).

It'd be nice if we could have the old behaviour back as an option `:graceful`), which can be described as:

> When the max redirects is hit, stop performing redirects and return the redirect response as is

Sample payload:

``` clj
{
 :status 302
 :location "http://some.location4"
 :trace-redirects [
    "http://some.location1", 
    "http://some.location2",
    "http://some.location3"
 ]
}
```

I think this is possible by reifying our own custom `RedirectStrategy` to comply with this behaviour:

I've tried to implement a custom ResponseInterceptor to update the `HttpContext` but it's not updated on each interceptor chain run so this needs to be implemented as a custom strategy (and not as elegant compared to a custom redirect strategy)

See https://github.com/apache/httpclient/blob/4.5.2/httpclient/src/main/java/org/apache/http/impl/execchain/RedirectExec.java#L113-L114

**Comment by rymndhng** [comment]
Another related idea would be to allow `:redirect-strategies` to allow users to pass in their own concrete RedirectStrategy if the argument implements `RedirectStrategy`

**Comment by dakrone** [comment]
@rymndhng It sounds like adding the option to go back to the original behavior is a good compromise, with the custom `RedirectStrategy` being the more advanced option. Are you interested in tackling either of these?

**Comment by rymndhng** [comment]
@dakrone Yep I'll give it a go!

**Comment by dakrone** [comment]
Raymond Huang writes:

> @dakrone Yep I'll give it a go!

Awesome! Looking forward to it!

--
;; Lee

**Comment by dakrone** [comment]
Closed by #326 and released in 3.5.0
